### PR TITLE
Respect PGP body hashes signed into sigchain

### PR DESCRIPTION
**Pull request by Sidnicious**
cc @maxtaco @oconnor663

### CODE REVIEW

**Comment by Sidnicious** [comment]
Fixes #6

File: src/sigchain/sigchain.iced [hunk 1]
```diff
@@ -36,47 +53,62 @@ exports.ParsedKeys = ParsedKeys = class ParsedKeys
         # We also merge userids to help prove key ownership.
```
Comment (oconnor663) [inline src/sigchain/sigchain.iced]: Can we update/move this comment to more fully describe what we're doing now?

File: src/sigchain/sigchain.iced [hunk 2]
```diff
@@ -36,47 +53,62 @@ exports.ParsedKeys = ParsedKeys = class ParsedKeys
...
-    cb null, parsed_keys, default_eldest_kid_for_testing
+  get_merged_pgp_key_manager : (kid) ->
+    @_kids_to_merged_pgp_key_managers[kid]
 
-  constructor : ({kids_to_key_managers}) ->
-    # Callers should use this class to get KeyManagers for KIDs they already
-    # have, but callers MUST NOT iterate over the set of KIDs here as though it
-    # were the valid set for a given user. The set of KIDs is *untrusted*
-    # because it comes from the server. We keep the map private to prevent that
-    # mistake.
```
Comment (oconnor663) [inline src/sigchain/sigchain.iced]: We should preserve this comment in some form.

File: src/sigchain/sigchain.iced [hunk 3]
```diff
@@ -36,47 +53,62 @@ exports.ParsedKeys = ParsedKeys = class ParsedKeys
...
-    @_kids_to_key_managers = kids_to_key_managers
+  get_pgp_key_manager_with_hash : (kid, hash) ->
+    @_kids_to_pgp_key_managers_by_hash[kid]?[hash]
 
-  get_key_manager : (kid) ->
-    @_kids_to_key_managers[kid]
+  get_nacl_key_manager : (kid) ->
+    @_kids_to_nacl_keys[kid]
+
+class KeyState
```
Comment (oconnor663) [inline src/sigchain/sigchain.iced]: Let's describe what this class is doing. (Sorry for all the comments about comments.)

File: src/sigchain/sigchain.iced [hunk 4]
```diff
@@ -124,21 +156,26 @@ class ChainLink
...
+  @_check_reverse_signatures : ({payload, key_state}, cb) ->
     esc = make_esc cb, "ChainLink._check_reverse_signatures"
     if payload.body.sibkey?
       kid = payload.body.sibkey.kid
-      key_manager = parsed_keys.get_key_manager kid
-      if not key_manager?
+      full_hash = payload.body.sibkey.full_hash
+      sibkey_key_manager = if full_hash?
+        # key_state hasn't been updated with the new sibkey's full hash yet
+        key_state.parsed_keys.get_pgp_key_manager_with_hash kid, full_hash
```
Comment (oconnor663) [inline src/sigchain/sigchain.iced]: Can we add a `TODO` here? In theory we should be testing the case where a merged version of the sibkey was valid, but the full hash version at the time of delegation technically was not.

**Comment by oconnor663** [comment]
Other than all my comment requests above, this looks good to me. @maxtaco do you care to take a look, or should we just :shipit:?

**Comment by maxtaco** [comment]
I'm good if you're good!

File: src/sigchain/sigchain.iced (hunk 2 above)
Comment (Sidnicious) [inline src/sigchain/sigchain.iced]: It's preserved above!

File: src/sigchain/sigchain.iced [hunk 5]
```diff
@@ -19,15 +19,32 @@ exports.ParsedKeys = ParsedKeys = class ParsedKeys
...
+      await parsed_keys._add_key {key_manager}, esc defer()
+      default_eldest_kid_for_testing or= key_manager.get_ekid().toString "hex"
+    cb null, parsed_keys, default_eldest_kid_for_testing
+
+  constructor : ->
+    # Callers should use this class to get KeyManagers for KIDs they already
+    # have, but callers MUST NOT iterate over the set of KIDs here as though it
+    # were the valid set for a given user. The set of KIDs is *untrusted*
+    # because it comes from the server. We keep the map private to prevent that
+    # mistake.
```
Comment (Sidnicious) [inline src/sigchain/sigchain.iced]: :point_up: here

File: src/sigchain/sigchain.iced (hunk 4 above)
Comment (Sidnicious) [inline src/sigchain/sigchain.iced]: Not sure I follow. We validate the reverse sig using the KM specified by `full_hash`, so we fail even if the sig would've been valid with the merged version.

**Comment by Sidnicious** [comment]
@oconnor663 I made most of these changes but I'm still unclear on one of your questions.

**Comment by Sidnicious** [comment]
@oconnor663 I think I'm going to merge this, since it's just a comment change. Let's chat once you're back online.

**Pull request merged**